### PR TITLE
segment_ids for text_to_lm_training_input outputs

### PR DESCRIPTION
**Pull request by changlan**
We add two optional tensors `input_segment_ids` and `input_positions` to the output of `text_to_lm_training_input`, to denote the segments in the sequence and the positions within each segment.

### CODE REVIEW

File: axlearn/common/input_lm_test.py [hunk 1]
```diff
@@ -185,78 +134,169 @@ def test_training_lm_processor_single_example(self, text: str):
...
+        #      2,     29,  3155,     1, 21820,   296,     2,    29,  3155]
+        window_size = 3
+
+        # Pad the concatenated sequence to 20 tokens.
+        # [    1,  21820,   296,     2,    29,  3155,     1, 21820,  8114,    2
+        #     29,  3155,     1, 21820,   296,     2,    29,  3155,      0,    0]
+        batch_size, max_len = 2, 5
+
+        # Always pad.
+        max_padding_fraction = 1.0
```
Comment (ruomingp) [inline axlearn/common/input_lm_test.py:231]: With this change, do we still test the truncation behavior?

File: axlearn/common/input_lm.py [hunk 2]
```diff
@@ -142,12 +163,45 @@ def process_batched(inputs: dict[str, Any]) -> dict[str, Any]:
...
+            segment_ids = tf.where(tf.equal(batched_input_ids, vocab.pad_id), 0, segment_ids)
+
+            assert segment_ids.shape[1] == max_len
+
+            def cummax_on_last_axis(x):
+                """tf.scan implementation of cummax."""
+                assert len(x.shape) == 2
+                # transpose to [seq, batch] for tf.scan, which iterates on dimension 0.
+                x = tf.transpose(x)
+                x = tf.scan(tf.maximum, x, initializer=tf.reduce_min(x, axis=0))
```
Comment (ruomingp) [inline axlearn/common/input_lm.py:186]: Neat use of tf.scan!

**Review by ruomingp — COMMENTED** [review]
Can you also add the PR description? Thanks!

File: axlearn/common/input_lm_test.py (hunk 1 above)
Comment (changlan) [reply axlearn/common/input_lm_test.py]: Great catch! I added a test case of the truncation behavior. PTAL

**Review by ruomingp — APPROVED** [review]

**Pull request merged**